### DISCUSSION TITLE
How to change grid layout based on odd or even page?

### DISCUSSION THREAD

**Discussion by remggo** [typst/typst#3409] [Q&A] [answered]
I have the use case, that I want to have a `#grid(columns: (1fr, 2fr), ...)` Layout in my document, but it needs to be changed based on even or odd page numers, so that in the layout the `1fr` column is always on the outside of the page.

The end result should look like this (ignore the width/height, this is just that the screenshots are smaller):

```typst
#set par(justify: true)

#set page(width: 10cm, height: 10cm, margin: (inside: 3cm), numbering: "1")

= Odd Page
#grid(columns: (2fr, 1fr), column-gutter: 1em, [#lorem(40)], [#text(blue, lorem(20))])

#pagebreak()
= Even Page
#grid(columns: (1fr, 2fr), column-gutter: 1em, [#text(blue, lorem(20))], [#lorem(40)])
```

![image](https://github.com/typst/typst/assets/3630154/32356516-5a5e-48da-9043-8ddafd07791f)

One thing that I also do not know how this should be possible with typst (if at all at the moment) is the overflow of text on the next page. The `margin: (inside: 3cm)` works as expected and this makes my inner margins correct on even and odd pages. However, the columns would ideally be swapped on the even/odd page so that the small column is on the outside of the page.

![image](https://github.com/typst/typst/assets/3630154/30984bab-3f3b-4fd0-9007-e503c91489ef)

Does anybody have any idea how to implement something like this? I cannot find any solution that works with 

`#set grid(columns: locate(loc => {..}))`

Or maybe this is not yet implemented?

**Comment by emilyyyylime** [comment]
Unfortunately the result of locate cannot be used as a value. You will instead need something like
```typ
#show grid: it => locate(loc => {
  if calc.mod(loc.page, 2) == 0 {
    set grid(columns: (2em, 1em)
    it
  } else etc..
})
```
But really I would write this as some separate function rather than a show rule

**Reply by remggo** [reply]
This works for one-page layouts and is already a big help. Yes putting it in a function will make it much more clean.

As long as the content of the columns of the grid fit on a single page, it will work fine (like here):
![image](https://github.com/typst/typst/assets/3630154/1acfcc12-3b9c-40b8-be48-bd7ab2823210)

But once the content of the columns do not fit anymore this will happen (instead of adapting correctly, black text inside, blue text outside):
![image](https://github.com/typst/typst/assets/3630154/68cdf141-d4c9-4ae2-8fab-f0b96591f864)

You do not happen to know how to tell typst, that there is a switch in position if the content of the grid should "overflow" to the next page?

**Reply by xkevio** [reply]
just a nitpick: its `calc.rem` but also `calc.even` and `calc.odd` exist, so no need to check it manually.

**Reply by emilyyyylime** [reply, marked as answer]
@remggo so, this is possible, but only by reimplementing text layouting. If you want to pursue it a good approach might be to locate/measure each paragraph/sentence of the gutters to see if it overflows the current page's gutter and move it to the next somehow if so

**Reply by remggo** [reply]
I fear this is what it comes down to :/ Thanks for the input.
